presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,472 @@
+                Optional<ColumnMapping> columnMapping = toPrestoType(session, connection, typeHandle);
+                log.debug("Mapping data type of '%s' column '%s': %s mapped to %s", tableHandle.getSchemaTableName(),
+                        columnName, typeHandle, columnMapping);
+                // skip unsupported column types
+                if (columnMapping.isPresent()) {
+                    boolean nullable = (resultSet.getInt("NULLABLE") != columnNoNulls);
+                    Optional<String> comment = Optional.ofNullable(resultSet.getString("REMARKS"));
+                    columns.add(JdbcColumnHandle.builder().setColumnName(columnName).setJdbcTypeHandle(typeHandle)
+                            .setColumnType(columnMapping.get().getType()).setNullable(nullable).setComment(comment)
+                            .build());
+                }
+                if (columnMapping.isEmpty()) {
+                    UnsupportedTypeHandling unsupportedTypeHandling = getUnsupportedTypeHandling(session);
+                    verify(unsupportedTypeHandling == IGNORE,
+                            "Unsupported type handling is set to %s, but toPrestoType() returned empty",
+                            unsupportedTypeHandling);
+                }
+            }
+            if (columns.isEmpty()) {
+                // In rare cases a table might have no columns.
+                throw new TableNotFoundException(tableHandle.getSchemaTableName());
+            }
+            return ImmutableList.copyOf(columns);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        if (BOOLEAN.equals(type)) {
+            return WriteMapping.booleanMapping("UInt8", booleanWriteFunction());
+        }
+        if (REAL.equals(type)) {
+            return WriteMapping.longMapping("Float32", realWriteFunction());
+        }
+        if (TIMESTAMP_MILLIS.equals(type)) {
+            return WriteMapping.longMapping("DateTime", timestampWriteFunction(TIMESTAMP_MILLIS));
+        }
+        if (VARBINARY.equals(type)) {
+            return WriteMapping.sliceMapping("String", varbinaryWriteFunction());
+        }
+        if (DOUBLE.equals(type)) {
+            return WriteMapping.doubleMapping("Float64", doubleWriteFunction());
+        }
+        if (type.getBaseName().equals("double")) {
+            return WriteMapping.doubleMapping("Float64", doubleWriteFunction());
+        }
+        if (VARCHAR.equals(type)) {
+            return WriteMapping.sliceMapping("String", varcharWriteFunction());
+        }
+        if (type.getBaseName().equals("char")) {
+            return WriteMapping.sliceMapping("String", varcharWriteFunction());
+        }
+        return legacyToWriteMapping(session, type);
```

Comment:
`legacyToWriteMapping` is deprecated & meant to be used by legacy code only. No need to use it in new code.

```suggestion
        throw new PrestoException(NOT_SUPPORTED, "Unsupported column type: " + type.getDisplayName());
```